Firestore/core/src/firebase/firestore/api/document_snapshot.h

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_API_DOCUMENT_SNAPSHOT_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_API_DOCUMENT_SNAPSHOT_H_
+
+#if !defined(__OBJC__)
+#error "This header only supports Objective-C++"
+#endif  // !defined(__OBJC__)
+
+#import <Foundation/Foundation.h>
+
+#include <string>
+#include <utility>
+
+#import "Firestore/Source/Model/FSTFieldValue.h"
+
+#include "Firestore/core/src/firebase/firestore/model/document_key.h"
+#include "Firestore/core/src/firebase/firestore/model/field_path.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@class FIRFirestore;
+@class FIRSnapshotMetadata;
+@class FIRDocumentReference;
+@class FSTDocument;
+
+namespace firebase {
+namespace firestore {
+namespace api {
+
+class DocumentSnapshot {
+ public:
+  DocumentSnapshot() = default;
+
+  DocumentSnapshot(FIRFirestore* firestore,
+                   model::DocumentKey document_key,
+                   FSTDocument* _Nullable document,
+                   bool from_cache,
+                   bool has_pending_writes)
+      : firestore_{firestore},
+        internal_key_{std::move(document_key)},
+        internal_document_{document},
+        from_cache_{from_cache},
+        has_pending_writes_{has_pending_writes} {
+  }
+
+  size_t Hash() const;
+
+  bool Exists() const {
```

Comment:
Wouldn't the style guide call for these to be `exists()`, `internal_document()`, and `document_id()`?

Reply:
Done.

I think it's optional in the style guide. I personally don't really like it, because it looks inconsistent, and also because it seems to create a contract of sorts (presumably if the member function stops being a simple getter, it would have to be renamed).